### PR TITLE
Add support for artists and albumartists multi-valued tags

### PR DESCRIPTION
**Pull request by jmbannon**
## Description
https://github.com/beetbox/beets/issues/505

Piggy-backs on the work done in https://github.com/beetbox/beets/pull/4582 to add support for:
- `artists`
- `artists_sort` (mediafile support added in https://github.com/beetbox/mediafile/pull/69)
- `artists_credit` (mediafile support added)
- `albumartists`
- `albumartists_sort` (mediafile support added)
- `albumartists_credit` (mediafile support added)
- `mb_artistids`
- `mb_albumartistids`

multi-tags as part of the default behavior when fetching tags from musicbrainz. The ones marked accordingly require mediafile >=0.12.0 to write the tags to the file. `setup.py` is updated to reflect this.

I think `albumtypes` has paved the way on how multi tags should be handled - using a delimiter and storing them in the database as a string, but writing them to tags as a multi (list).

Some considerations with this approach are:

- It requires no massive rewrite/migration/etc, falls in line with the 'KISS' database schema that beets has adopted
- Searching multi-fields will require regex (i.e. `artists::Eric Clapton`) since the db's literal value will be `Eric Clapton\␀B.B. King`
- mediafile abstracts whether it's actually written as a multi, or with a separator when using id3v23
  - Right now it uses `/` for a delimiter, example above would write `Eric Clapton/B.B. King`

Potential follow-ups:
- Add new mediafile field definitions for the ones marked above (https://github.com/beetbox/mediafile/blob/master/mediafile.py#L1785)
- Treat the `genre`/`genres` tag like this (PR is up! https://github.com/beetbox/beets/pull/4751)
- Introduce a way to read/write multi-fields as multi. From the CLI, specifying "Eric Clapton\␀B.B. King"  looks incredibly ugly.
  - See syntax discussion here: https://github.com/beetbox/beets/discussions/4792#discussioncomment-5919115
  - New syntax would translate to regex under the hood
- Add composer, performers, and other classical multi tags
- Configurable delimiter to use within mediafile
  - Low priority for me, it's time to move into the 21st century with multi-tags :smile: 

## To Do

- [x] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [x] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- [X] Tests. (Encouraged but not strictly required.)

### CODE REVIEW

File: beets/autotag/hooks.py
Comment (jmbannon) [inline beets/autotag/hooks.py:122]: This was missing from the other PR

**Comment by arsaboo** [comment]
Oh my God... this would be amazing. Can you add basic docs for someone to test this out? I'll test this out and report findings.

**Comment by jmbannon** [comment]
> Oh my God... this would be amazing. Can you add basic docs for someone to test this out? I'll test this out and report findings.

I haven't tested this yet aside from unit tests. I too will give it a spin when I get more free time.

**Comment by jmbannon** [comment]
I imported [this album](https://musicbrainz.org/release-group/75e7898c-393f-4159-9ad9-57966d9339e8) on my branch and verified that `albumartists` are indeed writing multi tags :tada: 

As for testing docs @arsaboo , multi tags for artists/albumartists should work out of the box - no additional configuration needed.

I'll add some more unit tests to make sure id3v23 works as expected

File: beets/autotag/mb.py
Comment (jmbannon) [inline beets/autotag/mb.py:281]: might be worth rewriting _multi_artist_credit and _flatten_credit_artist to only scan the artist-credit once. Currently, it's doing two passes

**Comment by jmbannon** [comment]
Any thoughts on this @sampsyo ? Would love some feedback on this - and sorry for being impatient 😅

File: beets/autotag/mb.py
Comment (sampsyo) [inline beets/autotag/mb.py:218]: Really nice work on this refactoring. Clever and also simple!

**Review by sampsyo — APPROVED** [review]
Wow!!! This is incredibly great work—many commendations, @jmbannon. I am super impressed by how simple these changes ended up being, especially given how low the bug number is for #505. :smiley: I'm sure we could find an even more ambitious, complex way to approach this, but this seems like a very reasonable way to get most of the benefit without changing everything all at once.

I had a couple of very tiny comments within, but if this seems to work for everybody and passes CI, I think it's basically good to go. Again, huge thanks to @jmbannon for tackling a seemingly-intractable problem with aplomb.

File: beets/autotag/hooks.py
Comment (jmbannon) [reply beets/autotag/hooks.py:78]: As boiler-plate as it is, I prefer it lol. For me, it helps to know which native field types are expected

**Comment by jmbannon** [comment]
> I am super impressed by how simple these changes ended up being, especially given how low the bug number is for #505 
I too was humbled by how simple this was. Goes to show I'm standing on the shoulders of giants 

One test failed because python <=3.8 and/or sqlite doesn't like the literal null terminator as a DB value. Changed it to `\␀`. For future work like adding a `contains` search syntax for multi's, this will be the delimiter used to create the underlying regex. Mediafile won't actually use this separator since the values get passed in as a list.

**Comment by JOJ0** [comment]
Approved test run. HTH.

**Comment by jmbannon** [comment]
Friendly bump - if everything looks good, hoping we can get this merged for more follow-up PRs ☺️ 
cc @JOJ0 @sampsyo

**Comment by JOJ0** [comment]
> Friendly bump - if everything looks good, hoping we can get this merged for more follow-up PRs relaxed cc @JOJ0 @sampsyo

Hi, I'm sorry, I might not be the best person to review here. I didn't really follow and am not sure if I want to review and make a call here, just not enough time. Just one thing I notice and am curious about: How is your change to albumtypes related to the changes to albumtypes in the recently merged PR #4582

**Comment by jmbannon** [comment]
> How is your change to albumtypes related to the changes to albumtypes in the recently merged PR #4582

So it uses the same exact mechanisms as albumtype/albumtypes. If you see in this bit of [mediafile code](https://github.com/beetbox/mediafile/blob/master/mediafile.py#L1954), `albumtype` and `albumtypes` are the same underlying field, but can be set in two different ways (single vs multi as a list).

The `mb_*` fields are this way. However, artists + albumartists have their own tag versus sharing one ([as seen here](https://github.com/beetbox/mediafile/blob/masexactter/mediafile.py#L1799)).

> Just a quick tutorial writeup as a comment in here, describing how to change some fields, how it affects the database entries, and so on. 

That's the thing - setting multi fields via CLI is not well-supported from this diff. The separator used in the DB is `\␀`, so to set albumartists to Eric Clapton + B.B.King, the value on the CLI would be `Eric Clapton\␀B.B. King``. Follow-up work is needed to introduce a slick way to set it via CLI. For querying this field, regex is needed unless you incorporate `\␀` in the regex itself.

That's not to take away from what this PR actually introduces - which is reading and setting the multi-tags from MusicBrainz onto files by default.

> Can you do me a favor and post a `python -m unittest ....` command to run them separately. Sorry for being lazy!! :-/

I can when I find time, but IMO the unit tests should suffice in showing how multi-tags work when interacting with the CLI. See [this one](https://github.com/beetbox/beets/pull/4743/files#diff-d8cb9eac8c4152a3f6cbc5fea274917a3869d2b48c1b168439342f5722b93a71R113) and [this other one](https://github.com/beetbox/beets/pull/4743/files#diff-875a94da4bd05e0aa5b96d04a4f9e3b92684c91695985e2151ae5cb1b22c0a2cR220) in particular. For now, querying multi-tags will require using regex.

And a note on the `\␀` separator, I prefer it over `; ` because I'm convinced there are some niche artists that exist who have a semi-colon in their name. I think it's best to avoid using that.

**Comment by jmbannon** [comment]
Additional thoughts on the above @sampsyo ?

**Comment by catduckgnaf** [comment]
This is amazing, can we get this merged?

**Comment by OxygenCobalt** [comment]
Finally got to test this @jmbannon. The implementation seems good except for the poor tag support (The same footguns with single-value MBIDs and "+" separators remain). I would definitely like to see another PR for adding ID3v2.4.0 null-value separators & separator customization, which I could help with.

**Comment by jmbannon** [comment]
> I would definitely like to see another PR for adding ID3v2.4.0 null-value separators & separator customization

For reading/writing multi values explicitly, definitely. Some discussion should probably be had on what the syntax of that will look like.

For separator customization, you can already set in the beets config to use id3v2.3, which will use a semi colon. TBH I don't think any more work is needed for building that out. I plan to focus exclusively on id3v2.4 multi support

**Comment by OxygenCobalt** [comment]
> For reading/writing multi values explicitly, definitely. Some discussion should probably be had on what the syntax of that will look like.

ID3v2.4 formats multiple values as just `foo\0bar\0baz` for multiple values @jmbannon.

**Comment by OxygenCobalt** [comment]
Now If you're talking about how to resolve the "querying artists with regex" issue @jmbannon, I came up with this syntax awhile back on a tagger I wrote before discovering beets:

- `foo | [...foo...]` = Match any instance of foo
- `[foo...]` = Match foo at the beginning
- `[...foo]` = Match foo at the end
- `[foo,bar,baz]` = Match exactly foo,bar,baz
- `[...foo,bar,baz...]` = Match foo,bar,baz in that order, but in any position
- `[...foo,bar,baz]` = Match foo,bar,baz in that order, but at the end
- `[foo,bar,baz...]` = Match foo,bar,baz in that order, but at the beginning

Assignment could be something like `field=foo` or `field=[foo,bar,baz]`.

**Comment by jmbannon** [comment]
> AFAIK ID3v2.4 formats multiple values as just `foo\0bar\0baz` @jmbannon.

This PR writes proper id3v2.4 separators on the audio files, not in the database though, since the null value isn't supported in older versions of python/sqllite

**Comment by jmbannon** [comment]
> 



> Now, If you're talking about how to resolve the "querying artists with regex" issue @jmbannon, I came up with this syntax awhile back on a CLI tagger I worked on before discovering beets:
> 
> * `foo | [...foo...]` = Match any instance of foo
> * `[foo...]` = Match foo at the beginning
> * `[...foo]` = Match foo at the end
> * `[foo,bar,baz]` = Match exactly foo,bar,baz
> * `[...foo,bar,baz...]` = Match foo,bar,baz in that order, but in any position
> * `[...foo,bar,baz]` = Match foo,bar,baz in that order, but at the end
> * `[foo,bar,baz...]` = Match foo,bar,baz in that order, but at the beginning
> 
> Assignment could be something like `field=foo` or `field=[foo,bar,baz]`.

This is a great start. One piece I'm not seeing is how would you query foo,bar,baz and not care about any order? Or do a contains_any query?

**Comment by OxygenCobalt** [comment]
> This PR writes proper id3v2.4 separators on the audio files, not in the database though, since the null value isn't supported in older versions of python/sqllite

Odd @jmbannon, because I tried the branch (Specifically did [Her Revolution / His Rope](https://musicbrainz.org/release/7aff5d94-effc-477a-aae2-6802974cd1c6)), but I kept getting "Burial + Four Tet + Thom Yorke" instead of correct null-terminated values. The `artists` tags were correctly populated.

> This is a great start. One piece I'm not seeing is how would you query foo,bar,baz and not care about any order? Or do a contains_any query?

I propose either `foo,bar,baz` or `[foo...bar...baz]` for this. Former is nicer, but of course harder to specify an artist if it has a comma.

**Comment by jmbannon** [comment]
@OxygenCobalt which field had the +s?

**Comment by OxygenCobalt** [comment]
Wait... I see. Looks like only the "native" artist fields like `TPE1` has +s. Non-standard fields like `ARTISTS` have them. I'm not really sure why this is the case @jmbannon, as `TPE1` & friends should be capable of supporting them. I'm assuming you're attempting to maximize compatibility in this regard?

**Comment by jmbannon** [comment]
> Wait... I see. Looks like only the "native" artist fields like `TPE1` has +s. Non-standard fields like `ARTISTS` have them. I'm not really sure why this is the case @jmbannon, as `TPE1` & friends should be capable of supporting them. I'm assuming you're attempting to maximize compatibility in this regard?

Correct. IMO artist should be treated like the display artist for a few reasons:
- compatibility
- would need a new field to preserve the artists and 'join fields' (+ is the join field in this case)
- mediafile, the underlying tag writer package, already has 'artists' implemented as the artist multi field

**Comment by OxygenCobalt** [comment]
> Correct. IMO artist should be treated like the display artist for a few reasons:

Okay, thanks for letting me know about that. I remembered that Auxio does prefer `ARTISTS` (and similar) by default over `TPE1` for this exact reason, so these files should work fine. This PR LGTM overall @jmbannon.

**Comment by OxygenCobalt** [comment]
Also, speaking further of querying multi-value without any order @jmbannon, I think something like `[foo/bar/baz]` would probably be the best option. No conflicts with bash syntax, consistent with other multi-value queries, and it can be chained for things like `[...foo/bar/baz...]` (any order any position).

**Comment by jmbannon** [comment]
> Also, speaking further of querying multi-value without any order @jmbannon, I think something like `[foo/bar/baz]` would probably be the best option. No conflicts with bash syntax, consistent with other multi-value queries, and it can be chained for things like `[...foo/bar/baz...]` (any order any position).

I like that but AC/DC would be too hard to query. Let's move this conversation into a GH discussion

**Comment by OxygenCobalt** [comment]
I think I've found a bug when trying to tag the [Paraglow](https://musicbrainz.org/release/9b3b0695-0cdc-4560-8486-8deadee136cb). If Picard is used, ALBUMARTIST is `Parannoul\0Asian Glow`. With Beets, ALBUMARTISTS is just `Parannoul`. Care to investigate this @jmbannon?

**Comment by jmbannon** [comment]
@OxygenCobalt Huge thanks for thoroughly testing this and finding that bug! Embarrassingly, I didn't update the autotagger code to use AlbumInfo to populate multi album-related fields :sweat_smile:  Updated and included exhaustive unit tests.

Could you try again using the latest changes?

**Comment by jmbannon** [comment]
Finally got around to using this in my new 'prod' tagging setup, all seems to be working as expected. Not having to manually use Picard for multi-tags is a game-changer. Looking forward to not having to use my fork :blush:

**Comment by jmbannon** [comment]
@OxygenCobalt Looked into it, and the reason is there is no `mediafile` definitions for those fields (https://github.com/beetbox/mediafile/blob/master/mediafile.py#L1785)

There is tests in `test_autotag.py` that shows they're on a mediafile item, but I'm guessing the write doesn't actually happen because of the above. I'll update the PR description reflect that

**Comment by OxygenCobalt** [comment]
So, it's not going to be implemented in this PR? That feels somewhat inconsistent @jmbannon.

**Comment by jmbannon** [comment]
@OxygenCobalt It's a different repo so it'd need its own PR, then a new release, then update beets to use said new release. Happy to do that - I still think it's reasonable to keep these fields in this PR. There are many other fields that beets stores that do not get written into the file tags.

**Comment by catduckgnaf** [comment]
I vote we merge this request. When @jmbannon updates mediafile via a PR, a new request can be opened.
Once this is released, I am sure other feedback will come in, and @jmbannon seems committed to updating and maintaining it.

I don't think we should hold back the merge further, because a different software/repo does not support this PR.

**Comment by OxygenCobalt** [comment]
Oh, I didn't even realize that it was an external dependency. I get why a new PR is needed, then.

**Comment by jmbannon** [comment]
Went ahead and made the PR in mediafile to add the missing tags: https://github.com/beetbox/mediafile/pull/69

**Comment by jmbannon** [comment]
After discussion in https://github.com/beetbox/mediafile/pull/69, made the multi `*_credits` tags singular to `*_credit`

**Comment by jmbannon** [comment]
Updated `setup.py` to require the latest version of mediafile, now that https://github.com/beetbox/mediafile/pull/69 is part of it. This will ensure all tags outlined in the PR description will get written to the files themselves.

**Comment by jmbannon** [comment]
friendly bump :) cc @sampsyo @wisp3rwind @JOJ0 
FWIW I've been using this in my 'prod' setup now with great success

**Comment by JOJ0** [comment]
Hi @jmbannon thanks for the bump! What would help me to at least use these new features in my personal beets branch that is my personal "prod setup", would be if I could easily cherry-pick a couple of commits into that branch. Do you think it's possible for you to maybe reorder commits and then squash them together in fewer commits. I'm not a fan of squashing everything together into only one commit. Maybe useful chunks that even have commit messages that describe. I know it's a burden to do stuff like that but it would helpf so much if it wouldn't be just so much commits. 

Not sure if these parts of beets are too much terrority for me personally to really properly review them but at least it would help me to pick into my prod-beets and make use of it constantly in my day to day tasks.

**Comment by catduckgnaf** [comment]
I think that is counter productive. What is the issue with more commits. Clearly you like the feature and want to use for yourself. Is there any good reason we can't merge it now?

My understanding is there are no breaking changes, so can how about we tag this feature as beta, and merge it in to get more feedback?

Let's not let perfection be the enemy of progress.

**Comment by JOJ0** [comment]
@catduckgnaf please read again what I wrote. I'm trying help here!

**Comment by arsaboo** [comment]
@JOJ0 We should be able to use `pip install --upgrade --force-reinstall --no-deps git+https://github.com/jmbannon/beets.git@j/multi` to test this out?

**Comment by jmbannon** [comment]
PRs get squashed on merge so it will look sane then. I will make a separate branch with a single commit - there is value in PR commit history to see changes based on discussion

**Comment by jmbannon** [comment]
Alrighty, use this to test @JOJ0 : https://github.com/beetbox/beets/pull/4888

**Comment by jmbannon** [comment]
@JOJ0 it is true, this PR here has 2 commits: https://github.com/beetbox/beets/pull/4883 When you merge, it turns into a single commit on master, as seen by `sampsyo merged commit [0e5ade4] into beetbox:master [3 days ago]`

**Comment by JOJ0** [comment]
I'm not sure anymore why we are arguing about how git and github work....and why we are drifting away from a simple request I made to help the project and keep things tidy....

Sure there _is_ an option to squash all the commits together into one commit but you have to _use_ it / click it when merging. I don't know for sure if other people like Adrian or anyoneelse use that often. There might definitely be reasons to do so!

But in that particular case: No! Commits were not squashed. It is still two commits. A merge-commit **is not** squashing together and rebasing into one commit. It is something else! After a regular PR merge we would usually have "the orginal commits count" plus a single merge-commit.

Screenshot 2023-08-30 at 18 32 56

**Comment by jmbannon** [comment]
Ah you're right. Squashing can be done via this button - beets should consider using that for cleaner history going forward instead of merge commits

image

**Comment by jmbannon** [comment]
With that said, I still don't want to wipe the history in this PR because I've used it as reference. The simple workaround is use the other PR I made for your single-commit test, and do a squash merge when if/when it gets approved and merged into master

**Comment by catduckgnaf** [comment]
@JOJ0  I disagree with you, I think many people would LOVE the feature @jmbannon has added, bugs or not. 
He did what nobody could do in over a decade, its amazing. 

@sampsyo, @JOJ0  are you going to implement @jmbannon  suggest on Github? 
I know you complained about talking about Github, but he just showed you how to make it better instead of wasting everyones time and turning away great developers who want to contribute.

I appreciate this lesson and now very much understand why forks happen in this community, its a shame we can't all work together.

**Comment by arsaboo** [comment]
@jmbannon we can retain this PR for the sake of history. However, to help this move along, it may be a good idea to have a single commit in your new PR so that @JOJ0 can test it easily. Note that we are not loosing any functionality and the two PRs can be functionally equivalent. Depending on the feedback, we can decide the next steps. 

I'm also going to test this PR.

**Comment by jmbannon** [comment]
Maybe got lost in our debate 😂 this PR has a single commit of this entire PR to use to test: https://github.com/beetbox/beets/pull/4888

@arsaboo @JOJ0

**Comment by JOJ0** [comment]
> @JOJ0 I disagree with you, I think many people would LOVE the feature @jmbannon has added, bugs or not. He did what nobody could do in over a decade, its amazing.
> 
> @sampsyo, @JOJ0 are you going to implement @jmbannon suggest on Github? I know you complained about talking about Github, but he just showed you how to make it better instead of wasting everyones time and turning away great developers who want to contribute.
> 
> I appreciate this lesson and now very much understand why forks happen in this community, its a shame we can't all work together.

Dear @catduckgnaf, I want to be honest with you: Two days ago I was quite motivated to move on with this PR and help jmbannon move forward. But after our debate and in the end receiving the message I quoted above, I'm not at all motivated to move on with that topic. You are accusing me of several things: I'm wasting your time, I'm the reason that people walk away from the project and you are telling me "how it should be done".

Do you really think that helps anyone?

Well, we might have very different opinions on how Open Source projects should be handled, but certainly your opinion is valuabvle, BUT that does not mean you should throw them at us in that manner. Please go ahead and "politely" submit your suggestions into a fitting thread like that one: https://github.com/beetbox/beets/discussions/4558

**Comment by jmbannon** [comment]
@JOJ0 I appreciate the time you volunteer to push beets forward. I'll apologize on behalf of @catduckgnaf , the bottom-line is we're both eager to see this PR through and it's been a bit frustrating waiting for a few months. But it's important to remember that open source doesn't owe us anything, including fast reviews/approvals/etc from folks like you who spend time maintaining great software for free.

I'm still determined to do what needs to be done to get this feature pushed so everyone can enjoy its value. I hope we can continue collaborating - I would love to expand multi-tag functionality beyond this PR. That requires good terms and healthy collaboration, which I hope we can all commit to going forward.

**Comment by JOJ0** [comment]
No. Please dont. I never said I want a single commit. Please read again...

**Comment by catduckgnaf** [comment]
@JOJ0 I am sorry and wanted to apologize.
I simply want to see this great software become even better!

**Comment by JOJ0** [comment]
> @JOJ0 I am sorry and wanted to apologize.
> 
> I simply want to see this great software become even better! 

Apology accepted. :-) We will do that! Soonish!

**Comment by JOJ0** [comment]
I tried my best to go ahead and wrap my head around this feature. I read the comments history twice, the media file additions PR and did - at least - some playing around with it myself.

Thanks for testing [over there](https://github.com/beetbox/beets/pull/4888) @arsaboo, that helped a lot to make me understand some things without having to look at code or play around myself initially. Thanks to you both @arsaboo and @jmbannon!

Some playing around with a "two-artists-release"


Database:

```
$ beet info -l fracture binga drift | grep -e artists -e ids
        albumartists: Fracture\␀Sam Binga
 albumartists_credit: Fracture\␀Sam Binga
   albumartists_sort: Fracture\␀Binga, Sam
             artists: Fracture\␀Sam Binga
      artists_credit: Fracture\␀Sam Binga
         artists_ids:
        artists_sort: Fracture\␀Binga, Sam
   mb_albumartistids: 96999054-38f8-47da-8906-a6b763aa37ce\␀bfc15bf3-2904-4d2c-990b-5e2c350654b0
        mb_artistids: 96999054-38f8-47da-8906-a6b763aa37ce\␀bfc15bf3-2904-4d2c-990b-5e2c350654b
```

File tags via beets, prior to updating mediafile (still 0.11.0):

```
$ beet info fracture binga drift | grep -e artists -e ids
      albumartists: Fracture; Sam Binga
           artists: Fracture; Sam Binga
 mb_albumartistids: 96999054-38f8-47da-8906-a6b763aa37ce; bfc15bf3-2904-4d2c-990b-5e2c350654b0
      mb_artistids: 96999054-38f8-47da-8906-a6b763aa37ce; bfc15bf3-2904-4d2c-990b-5e2c350654b0
```

File tags via mutagen-inspect, prior to updating mediafile (still 0.11.0):

```
$ mutagen-inspect "/home/jojo/Documents/Sync/beets_test_music/Uncategorized/Fracture & Sam Binga - 2022 - Omura/10 - Fracture & Sam Binga - Omura - 6 Foot Drift (2022).mp3" | grep -i -e artists -e ids
TXXX=ALBUMARTISTS=Fracture / Sam Binga
TXXX=ALBUMARTISTSORT=Fracture & Binga, Sam
TXXX=ALBUM_ARTISTS=Fracture / Sam Binga
TXXX=ARTISTS=Fracture / Sam Binga
```


File tags via mutagen-inspect, prior to updating mediafile (still 0.11.0):

```
$ mutagen-inspect "/home/jojo/Documents/Sync/beets_test_music/Uncategorized/Fracture & Sam Binga - 2022 - Omura/10 - Fracture & Sam Binga - Omura - 6 Foot Drift (2022).mp3" | grep -i -e artists -e ids
TXXX=ALBUMARTISTS=Fracture / Sam Binga
TXXX=ALBUMARTISTSORT=Fracture & Binga, Sam
TXXX=ALBUM_ARTISTS=Fracture / Sam Binga
TXXX=ARTISTS=Fracture / Sam Binga
```

Updated mediafile to 0.12, writing tags:

```
$ beet write fracture binga drift
Fracture & Sam Binga - Omura - 10 - 6 Foot Drift (2022/XW)  [APHALP004] f:MP3-128kCBR a:70-01/ASIS |Electro, Metal, Drum And Bass
  artists_sort:  -> Fracture\␀Binga, Sam
  artists_credit:  -> Fracture\␀Sam Binga
  albumartists_sort:  -> Fracture\␀Binga, Sam
  albumartists_credit:  -> Fracture\␀Sam Binga
```


File tags beets now sees:

```
$ beet info fracture binga drift | grep -e artists -e ids
       albumartists: Fracture; Sam Binga
albumartists_credit: Fracture; Sam Binga
  albumartists_sort: Fracture; Binga, Sam
            artists: Fracture; Sam Binga
     artists_credit: Fracture; Sam Binga
       artists_sort: Fracture; Binga, Sam
  mb_albumartistids: 96999054-38f8-47da-8906-a6b763aa37ce; bfc15bf3-2904-4d2c-990b-5e2c350654b0
       mb_artistids: 96999054-38f8-47da-8906-a6b763aa37ce; bfc15bf3-2904-4d2c-990b-5e2c350654b0
```


File tags mutagen-inspect now sees:

```
$ mutagen-inspect "/home/jojo/Documents/Sync/beets_test_music/Uncategorized/Fracture & Sam Binga - 2022 - Omura/10 - Fracture & Sam Binga - Omura - 6 Foot Drift (2022).mp3" | grep -i -e artists -e ids
TXXX=ALBUMARTISTS=Fracture / Sam Binga
TXXX=ALBUMARTISTSORT=Fracture & Binga, Sam
TXXX=ALBUMARTISTS_CREDIT=Fracture / Sam Binga
TXXX=ALBUMARTISTS_SORT=Fracture / Binga, Sam
TXXX=ALBUM_ARTISTS=Fracture / Sam Binga
TXXX=ARTISTS=Fracture / Sam Binga
TXXX=ARTISTS_CREDIT=Fracture / Sam Binga
TXXX=ARTISTS_SORT=Fracture / Binga, Sam
```

First of all, is what I'm seeing here the expected outcome?

**Comment by JOJ0** [comment]
I'm very close to wanting to hit that merge button. Some questions come up:

- Where (in the code) is the replacement of the delimiter we save in the database with the one we finally save in the files tags happening? In this case it seems like a semicolon is used when saving (;).

- In your initial description of this PR you are suggesting some possible next steps. Which route would you like to go?

- Does the thumbs-up you put on my last message mean "Yes I'd like to take a minute and rebase some commits". In any case, I have another proposal what could be done to make that history survive (which I would prefer) but make it more valueable, with just a little work: Use `git rebase -i <commit_before_first_feature_commit> and then instead of using reordering and squashing, you could simply use the "reword" feature and rewrite some of those "single word" commit messages to short sentences. It would keep all your commits as-is, but help "future code reviewers" a lot when they stumble across those changes in the master branch eg. a couple of years. We will never know what might happen to github/gitlab/or wherever the future hosts code - but what will always remain with us because it's undesdructible - is the actual git history! It would be wonderful if single-word-commits could be just reworded. 

Some random examples from somewhere around beets, how a descriptive commit messages could look like:

```
    replaygain: better types to represent peak methods

    thanks @redacted_user in review https://github.com/beetbox/beets/pull/3996 for
    this suggestion to properly represent the sum type
```

```
    replaygain test: extend tests for skipping items

    Test more combinations of tags that might initially be present and
    expected tags. The R128 codepath and the case of having the wrong type
    of tags wasn't really tested before.
```

I know that not everyone is as tidy / likes to write a lot of words as the messages above do - but at least - 
We don't need exhaustive sentences here and not even more than a headline, but when I look through our git history it just helps so tremendously if I at least see "part of beets / plugin name: tiny headline what's it about"

I'm trying again to make this happen - if not - I'll give up ;-)

**Comment by jmbannon** [comment]
@JOJ0 Much thanks for all your time testing this out. Over the course of today, I'll try to respond to all of your questions from top-to-bottom

**Comment by jmbannon** [comment]
@JOJ0 All your commands look good. In the database, it writes `Fracture\␀Binga, Sam` to indicate usage of the id3v2.4 multi-tag separator.

I'm not familiar with mutagen-inspect - I assume ` / ` means it's separate with id3v2.4 null separator under-the-hood, especially since beets reports with `; ` as the 'human-readable' separator.

> Where (in the code) is the replacement of the delimiter we save in the database with the one we finally save in the files tags happening? In this case it seems like a semicolon is used when saving (;).

See https://github.com/beetbox/beets/blob/4f58b03be372b732e6a2059643e90fe3b473ab5a/beets/dbcore/types.py#L283 - it's a class that orchestrates list-based field representation. When we write to:
  - mediafile: the value itself will be in list form (passed within `beets/autotag/__init__.py`). The unit tests show the expected lists within the mediafile object, which indicates it supports id3v2.4 tagging.
  - database: Will write using the `to_sql` method in the DelimitedString class, which does a join using the separator. This is seen in your `update` command
  - beets info: It will see if the value is a list, and print them using .join("; "). This happens for any list-based field including `albumtypes` (which created the DelimitedString helper class). I forget where this happens, but I remember seeing it because I too was concerned it wasn't writing properly

> In your initial description of this PR you are suggesting some possible next steps. Which route would you like to go?
The ones on the top-of my mind are:
- more explicit printing within `info` indicating tags are multi
- better multi-search (as seen in the discussion linked in the description)

I suggest we do these in a follow-up PR since this one is already pretty big

>  I have another proposal what could be done to make that history survive 
I really think we should squash merge this as a single-commit to master since it's adding a single feature with tests to beets. My concern regarding history was just for the PR/branch. Future reviewers can refer to this PR to get the history. And worst-case if we need to revert the commit, it will be seamless since it's packaged into a single one.

> I see we have very little documentation on how to use this, but maybe it's too early to exhaustivle describe this feature. What's your plans?
IMO I don't think this PR needs any additional docs added to the readthedocs - it's simply adding more musicbrainz fields (which aren't documented as-is), and there already is a notion of using id3v2.3 vs id3v2.4 tags here: https://beets.readthedocs.io/en/stable/reference/config.html?highlight=id3#id3v23

> How is the decisions going around how this can be used from the cli without regex.
Nothing is in stone yet, but @sampsyo provided some good wisdom on how we should approach it. Again, I think it's okay to put this off for a future PR + reboot the discussion. Any multi-tag, regardless of id3v2.3 vs 2.4 has the same issue - it's not specific to this PR changeset.

**Comment by jmbannon** [comment]
>  how a descriptive commit messages could look like:

GitHub lets you change the commit message when you squash it. If you're okay with that approach, I can draft a message that you can use if/when you do the squash + merge

**Comment by JOJ0** [comment]
> >  how a descriptive commit messages could look like:
> 
> 
> 
> GitHub lets you change the commit message when you squash it. If you're okay with that approach, I can draft a message that you can use if/when you do the squash + merge

Alright, then lets go that way and I'll hit that mergedquas button :-)

**Comment by jmbannon** [comment]
@JOJ0 Sounds good! I'll get something drafted later today

**Comment by jmbannon** [comment]
Draft message for the squash merge commit:

```
multi-tags: add support for `artists`, `albumartists` related multi-tags (#4743)

  Adds the following fields with id3v2.4 multi-tag support to autotag:
    - artists
    - artists_sort (mediafile support added in multi-tags for artists + albumartists sort and credits mediafile#69)
    - artists_credit (mediafile support added)
    - albumartists
    - albumartists_sort
    - albumartists_credit
    - mb_artistids
    - mb_albumartistids
  MusicBrainz support to populate + write the above multi-tags by default. Can be toggled to use id3v2.3 or id3v2.4 tags via existing beets configuration.
  Big thanks to @JOJ0, @OxygenCobalt, @arsaboo for testing + @sampsyo for the initial code review 
```

**Comment by JOJ0** [comment]
@jmbannon I did some nitpicking on the wording in docs, changelog and PR/issue naming. Sometimes we referred as multi tags, multi-tags, multi-valued tags, multiple-valued tags, and I decided that mutli-valued tags is the most descriptive one which still is not too longish. Also in code we have MULTI_VALUE_DSV for example which fits with that decision.

**Comment by KingDuckZ** [comment]
I've been searching around for a way to split the new arrays, this is not working for me: `default: '%first{$albumartists}'` by the looks of it it's naming directories with the internal representation of the arrays, so with the nullchar. How can I get just the first entry?

**Comment by KingDuckZ** [comment]
While waiting I'm trying to import my music and noticed another issue with the rewrite plugin, it crashes beets if I try to alter albumartists with it. Looking at the backtrace, it tries to convert a list into lowercase. Just my guess but this change intoduced lists where previously there could only be strings?

**Pull request merged**
